### PR TITLE
Correctly validate joins between windowed and non-windowed sources

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Fixes: https://github.com/confluentinc/ksql/issues/4389

Previously, KSQL did nothing to validate a join between two sources had keys with compatible windowing.

* it did not check that both sources were either windowed or non-windowed.  Joining a non-windowed source to a windowed source will never result in any output as the binary keys can never match.
* it did not check that both sources had compatible window types. Session windows include both window start and end times in the serialized key, where as hopping and tumbling windows only serialize the window start. For this reason, session windowed sources can only be joined with other session windowed sources. Hopping and tumbling windowed sources can only be joined with other hopping and tumbling windowed sources.

This commit adds these checks and suitable QTT tests to prove them.

### Testing done 

test cases added to QTT

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java [hunk 1]
```diff
@@ -397,14 +387,95 @@ protected AstNode visitJoin(final Join node, final Void context) {
...
+      if (left.getDataSource().getName().equals(right.getDataSource().getName())) {
+        throw new KsqlException(
+            "Can not join '" + left.getDataSource().getName().toString(FormatOptions.noEscape())
+                + "' to '" + right.getDataSource().getName().toString(FormatOptions.noEscape())
+                + "': self joins are not yet supported."
+        );
+      }
+    }
+
+    private void throwOnIncompleteJoinCriteria(
```
Comment (vpapavas) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: This checks if the left and right alias are the same? Might want to consolidate it under the self join check. The current name is a bit misleading

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java [hunk 2]
```diff
@@ -397,14 +387,95 @@ protected AstNode visitJoin(final Join node, final Void context) {
...
+
+      if (leftWindowType.isPresent() != rightWindowType.isPresent()) {
+        throw windowedNonWindowedJoinException(left, right, leftWindowType, rightWindowType);
+      }
+
+      if (!leftWindowType.isPresent()) {
+        return;
+      }
+
+      final WindowType leftWt = leftWindowType.get();
```
Comment (vpapavas) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: Just to make sure: Session can only join with session whereas hopping and tumbling can join with each other interchangeably?

Comment (vpapavas) [reply ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: Nevermind, tests are self-explanatory :)

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: No, this checks the the two sides of the join criteria reference both sources, e.g. it rejects a join like:

```
... FROM A JOIN B ON B.COL1 = B.COL2;
```

Note: source `A` is not included in the join criteria.

**Pull request merged**